File: TDD/TDD Coding Practice String Transformer/Code/Python/dummy/original/validator.py
class ValidatingResult(object):
    VRFR_AVAIL_TRANS_NOT_SPECIFIED = 'validating_result_failed_reason_avail_trans_not_specified'
    VRFR_ADD_ALREADY_EXISTED_IN_CHAIN_TRANS = 'validating_result_failed_reason_add_already_existed_in_chain_trans'
    VRFR_CHAIN_TRANS_NOT_SPECIFIED = 'validating_result_failed_reason_chain_trans_not_specified'
    VRFR_CHAIN_EMPTY = 'validating_result_failed_reason_chain_empty'

    def __init__(self, is_succeeded, failed_reason=None):
        self.is_succeeded = is_succeeded
        self.failed_reason = failed_reason

    @staticmethod
    def succeeded_result():
        return ValidatingResult(True)

    @staticmethod
    def failed_result(failed_reason):
        return ValidatingResult(False, failed_reason)


class ParamValidatingRule(object):
    def __init__(self, param, failed_pred, failed_action, failed_reason):
        self.param = param
        self.failed_pred = failed_pred
        self.failed_action = failed_action
        self.failed_reason = failed_reason

class Validator(object):
    @staticmethod
    def validate_param(param_validating_rule):
        if param_validating_rule.failed_pred(param_validating_rule.param):
            param_validating_rule.failed_action()
            return ValidatingResult.failed_result(param_validating_rule.failed_reason)

        return ValidatingResult.succeeded_result()

    @staticmethod
    def validate(param_validating_rules):
        for rule in param_validating_rules:
            validating_result = Validator.validate_param(rule)
            if not validating_result.is_succeeded:
                return validating_result

        return ValidatingResult.succeeded_result()

File: TDD/TDD Coding Practice String Transformer/Code/Java/src/fayelab/tdd/stringtransformer/instruction/addall/Validator.java
package fayelab.tdd.stringtransformer.instruction.addall;

import java.util.List;
import java.util.function.Predicate;

import fayelab.tdd.stringtransformer.instruction.addall.ValidatingResult.FailedReason;

public class Validator
{
    public static ValidatingResult validate(List<ParamValidatingRule<?>> rules)
    {
        for(ParamValidatingRule<?> rule : rules)
        {
            ValidatingResult validatingResult = validateParam(rule);
            if(!validatingResult.isSucceeded())
            {
                return validatingResult;
            }
        }

        return ValidatingResult.succeededResult();
    }

    private static <T> ValidatingResult validateParam(ParamValidatingRule<T> rule)
    {
        if(rule.getFailedPred().test(rule.getParam()))
        {
            return ValidatingResult.failedResult(rule.getFailedReason());
        }

        return ValidatingResult.succeededResult();
    }
}

class ValidatingResult
{
    enum FailedReason
    {
        NONE,
        AVAIL_TRANS_NOT_SPECIFIED,
        ADD_ALREADY_EXISTED_IN_CHAIN_TRANS,
        CHAIN_TRANS_NOT_SPECIFIED,
        CHAIN_EMPTY,
        SOURCE_STR_EMPTY,
        SOURCE_STR_ILLEGAL
    }

    private boolean succeeded;
    private FailedReason failedReason;

    private ValidatingResult(boolean succeeded, FailedReason failedReason)
    {
        this.succeeded = succeeded;
        this.failedReason = failedReason;
    }

    public boolean isSucceeded()
    {
        return succeeded;
    }

    public FailedReason getFailedReason()
    {
        return failedReason;
    }

    public static ValidatingResult succeededResult()
    {
        return new ValidatingResult(true, FailedReason.NONE);
    }

    public static ValidatingResult failedResult(FailedReason failedReason)
    {
        return new ValidatingResult(false, failedReason);
    }
}

class ParamValidatingRule<T>
{
    private T param;
    private Predicate<T> failedPred;
    private FailedReason failedReason;

    static <T> ParamValidatingRule<T> paramValidatingRule(T param, Predicate<T> failedPred, FailedReason failedReason)
    {
        return new ParamValidatingRule<>(param, failedPred, failedReason);
    }

    private ParamValidatingRule(T param, Predicate<T> failedPred, FailedReason failedReason)
    {
        this.param = param;
        this.failedPred = failedPred;
        this.failedReason = failedReason;
    }

    public T getParam()
    {
        return param;
    }

    public Predicate<T> getFailedPred()
    {
        return failedPred;
    }

    public FailedReason getFailedReason()
    {
        return failedReason;
    }
}
